waterloo-rocketry/cansw_processor_stm
# Concurrent logging

maybe this speeds up logging idk

the concept is to split each log buffer into `n` sections, where each section will be handed off to tasks to write in. When a task uses `logInfo`, logger calculates the next free section and gives it to the task. No other task can have this section. Now multiple tasks can write to their own sections concurrently.

The downside is that most msgs do not fill the entire section, so there's a LOT of empty space. We should optimize the size of sections to be as small as possible. Possibly break up the VN log msgs into multiple msgs because that's constraining the section length requirement rn.

Other downside: buffers fill up a lot faster so it dumps to SD card more frequently. Probably not issue cuz SD card write speed is not a bottleneck.

-------------------------------------
things to test:
- run the program for a while
  - breakpoint somewhere in `log.c and verify these counters in `log.c` are all 0 
![image](https://github.com/user-attachments/assets/65be6a66-84e2-4194-bcec-0200f078db4c)
  - also breakpoint in `vn_handler.c` [here](https://github.com/waterloo-rocketry/cansw_processor_stm/pull/65/files#diff-dacc8aeaa93688d1ae35e8c37339d384c1f4792a133d041aff485948e8ccff92R133) to check the values of `diff1, diff2, pass1, pass2`
- open the log file from the sd card and verify every line logged is padded to the same alignment with `*` characters like the screenshot: 
![image](https://github.com/user-attachments/assets/9f901f26-884e-4a72-9413-8b6e90c03ad2)
